<a href="https://colab.research.google.com/github/MtalhaShzd/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MtalhaShzd/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

# 1. My Lane (or Freestyle) and Why

I have chosen the **Refresh / Content Opportunity Scoring** lane.

I selected this lane because it focuses on helping content teams identify which pages should be reviewed first for possible updates. Since organizations may have thousands of content pages, it is not practical to review every page manually. A machine learning model can rank pages using observable search performance and content metrics, allowing teams to prioritize their efforts more efficiently. I also chose this lane because the starter notebooks introduced this workflow, making it a good foundation for learning applied machine learning.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

# 2. The Question: Decision, Action, Cost of a Wrong Call

### Research Question

Can observable search performance and content signals be used to identify and rank pages that should be reviewed first for potential content refresh?

### Decision

The decision is which pages should be reviewed before others.

### Action

An SEO specialist or content manager can review the highest-ranked pages first and decide whether they need to be updated, optimized, or investigated further.

### Cost of a Wrong Recommendation

If the model ranks pages incorrectly, valuable pages that need attention may be missed, while time is spent reviewing pages that are already performing well. This reduces the efficiency of the content review process.

### Why Machine Learning?

Instead of relying on a single rule, machine learning can combine multiple observable signals such as impressions, CTR, content age, search position, and word count to support better prioritization decisions.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [7]:
# ============================================
# Quick Look at the Data
# ============================================

import os
import subprocess
import sys
import pandas as pd

REPO_DIR = "flyrank-ml-internship"
REPO_URL = "https://github.com/MtalhaShzd/flyrank-ml-internship.git"

# Clone your repository if it doesn't already exist
if not os.path.isdir(f"/content/{REPO_DIR}"):
    subprocess.run(
        ["git", "clone", REPO_URL],
        check=True
    )

# Change to repository root
os.chdir(f"/content/{REPO_DIR}")

print("Current Working Directory:", os.getcwd())

# Load the dataset
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("=" * 50)
print("Dataset Overview")
print("=" * 50)

print(f"Total Pages   : {len(df)}")
print(f"Total Columns : {len(df.columns)}")

# Discovery 1
corr_search = df["search_volume"].corr(df["impressions_90d"])

# Discovery 2
corr_position = df["avg_position"].corr(df["ctr"])

# Discovery 3
median_age = df.groupby("trend_direction")["content_age_days"].median()

print("\nCorrelation between search_volume and impressions_90d:")
print(round(corr_search, 3))

print("\nCorrelation between avg_position and CTR:")
print(round(corr_position, 3))

print("\nMedian Content Age by Trend Direction:")
print(median_age)

Current Working Directory: /content/flyrank-ml-internship
Dataset Overview
Total Pages   : 30000
Total Columns : 44

Correlation between search_volume and impressions_90d:
0.001

Correlation between avg_position and CTR:
-0.073

Median Content Age by Trend Direction:
trend_direction
down      216.0
flat      231.0
new       279.0
stable    300.0
up        291.5
Name: content_age_days, dtype: float64


# 3. Quick Look at the Data

The starter dataset contains **30,000 pages**, providing a realistic dataset for experimentation.

The correlation between **search volume** and **impressions** is approximately **0.001**, indicating that keyword search volume alone does not strongly explain page impressions.

The correlation between **average position** and **CTR** is approximately **-0.073**, showing only a very weak relationship.

The median content age also differs across trend groups. Stable pages have a higher median content age (300 days), while down pages have a lower median content age (216 days). These observations suggest that multiple signals may be useful when deciding which pages should be reviewed instead of relying on a single metric.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

# 4. Careful Words: What I Can and Can't Claim

This project aims to support content review decisions by identifying pages that may deserve attention based on observable search and content signals.

The results will describe observed patterns and directional relationships within the available dataset. They should not be interpreted as proof that updating a page will improve search performance or that the model can predict Google's ranking algorithm.

The recommendations produced by this project are intended to support human decision-making rather than replace it.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.